### PR TITLE
[SPARK-47497][SQL] Make `to_csv` support the output of `array/struct/map/binary` as pretty strings

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims make `to_csv` 
- support the output of `array/struct/map/binary` as `pretty strings`.
- not support `variant`.

### Why are the changes needed?
This PR was generated from follow-up comment suggestions https://github.com/apache/spark/pull/44665#issuecomment-2011239475,
image

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
- Update existed UT.
- Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
1.Why not use `ToPrettyString`?  
Because in `ToPrettyString`, the data output logic for `DateType` is `different from` that of csv itself, eg:
- `ToPrettyString`

- `to_csv`

File: sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala [hunk 1]
```diff
@@ -603,4 +605,118 @@ class CsvFunctionsSuite extends QueryTest with SharedSparkSession {
...
+      StructField("age", LongType),
+      StructField("name", StringType),
+      StructField("v", VariantType)))
+    val schema = StructType(Seq(
+      StructField("key", LongType),
+      StructField("value", valueSchema)))
+
+    val df = spark.createDataFrame(rows, schema)
+    val actual = df.select(to_csv($"value"))
+    checkAnswer(actual, Row("2,Alice,\"\""))
```
Comment (panbingkun) [inline sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:695]: Although this output seems a little weird, as `VariantVal` improves the method `toString` logic, it should be able to better display human-readable text.
https://github.com/apache/spark/blob/ca44489f458554714a620827b5fe2978a30db7f9/common/unsafe/src/main/java/org/apache/spark/unsafe/types/VariantVal.java#L101-L109

**Comment by panbingkun** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala [hunk 2]
```diff
@@ -36,9 +36,9 @@ class UnivocityGenerator(
   writerSettings.setHeaders(schema.fieldNames: _*)
   private val gen = new CsvWriter(writer, writerSettings)
 
-  // A `ValueConverter` is responsible for converting a value of an `InternalRow` to `String`.
+  // A `ValueConverter` is responsible for converting a value of an `Any` to `String`.
   // When the value is null, this converter should not be called.
-  private type ValueConverter = (InternalRow, Int) => String
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala:41]: `(InternalRow, Int)` is more efficient than `Any`, as we can avoid boxing for primitive values.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala:41]: I get that we want to recursively call the converter for array, then we can use `(SpecializedGetters, Int)`

File: sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:695]: I'm fine to ban variant type in `to_csv`, as it's new.

Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:695]: Okay

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala (hunk 2 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala:41]: Done.

File: python/pyspark/sql/functions/builtin.py [hunk 3]
```diff
@@ -15591,12 +15591,12 @@ def to_csv(col: "ColumnOrName", options: Optional[Dict[str, str]] = None) -> Col
...
-    +-----------------------+
-    |to_csv(value)          |
-    +-----------------------+
-    |2,Alice,"[100,200,300]"|
-    +-----------------------+
+    >>> df.select(sf.to_csv(df.value)).show(truncate=False)
+    +-------------------------+
+    |to_csv(value)            |
+    +-------------------------+
+    |2,Alice,"[100, 200, 300]"|
```
Comment (panbingkun) [inline python/pyspark/sql/functions/builtin.py:15598]: Aligning with `ToPrettyString`, adding a `space` between array elements seems reasonable, eg:
image

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala [hunk 4]
```diff
@@ -64,33 +65,143 @@ class UnivocityGenerator(
...
+          while (i < row.numFields) {
+            builder.append(", ")
+            if (row.isNullAt(i)) {
+              if (nullAsQuotedEmptyString) {
+                builder.append(options.nullValue)
+              } else {
+                builder.append(null.asInstanceOf[String])
+              }
+            } else {
+              builder.append(converters(i)(row, i))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala:195]: Can we add a function for the common null handling code?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala (hunk 4 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala:195]: Okay, I found a bug in the original (Before applying `this PR`) `to_csv`, as follows:
image
https://github.com/apache/spark/blob/227a50a1766ac1476b0031e1c60d2604eccdb9a7/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala#L100-L104
image

when `row.isNullAt(i)` && `nullAsQuotedEmptyString == false`, 
`values(i)` is not assigned, so it is the default value null, which seems incorrect?

image
According to this interpretation, we should assign value `""`.

@cloud-fan Should we fix it in this PR or open a separate PR to handle it?

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala:195]: Please let me check a possible issue before merging.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0.

File: sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala [hunk 5]
```diff
@@ -603,4 +605,167 @@ class CsvFunctionsSuite extends QueryTest with SharedSparkSession {
...
+    checkError(
+      exception = intercept[SparkUnsupportedOperationException] {
+        df.select(from_csv(lit("data"), valueSchema, Map.empty[String, String])).collect()
+      },
+      errorClass = "UNSUPPORTED_DATATYPE",
+      parameters = Map("typeName" -> "\"VARIANT\"")
+    )
+  }
+
+  test("SPARK-47497: the input of to_csv must be StructType") {
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:757]: What if the input is a nested structure? For example, the item of the Array is a Map, does it remain consistent with what we've seen before?

Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:757]: If it's already supported, should we add some corresponding test cases?

Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:757]: Let me add a UT for it.

Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:757]: @LuciferYang 
The related UT here: https://github.com/apache/spark/pull/45692

**Pull request closed without merging**